src/otel/src/lib.rs

```diff
@@ -0,0 +1,317 @@
+use opentelemetry::global::BoxedSpan;
+use opentelemetry::trace::{
+    Span, SpanBuilder, SpanKind, TraceId, Tracer, TracerProvider,
+};
+use opentelemetry::{global, KeyValue};
+use opentelemetry_otlp::Protocol;
+use opentelemetry_otlp::WithExportConfig;
+use opentelemetry_sdk::trace::{Config, BatchConfigBuilder, Sampler};
+use opentelemetry_sdk::{runtime, Resource};
+use std::ffi::{c_char, CStr, CString};
+use std::{ptr, time};
+use std::ptr::addr_of;
+use std::slice;
+use std::sync::{Arc, OnceLock};
+use tokio::sync::mpsc;
+use tokio::sync::mpsc::{Receiver, Sender};
+
+
+const TRACEPARENT_HEADER_LEN: u8 = 55;
+const TIMEOUT: time::Duration = std::time::Duration::from_secs(10);
+
+#[repr(C)]
+pub struct nxt_str_t {
+    pub length: usize,
+    pub start: *const u8,
+}
+
+// Stored sender channel to send spans or a shutdown message to within the
+// Tokio runtime.
+unsafe fn nxt_otel_rs_span_tx(destruct: bool) -> *const OnceLock<Sender<SpanMessage>> {
+    static mut SPAN_TX: OnceLock<Sender<SpanMessage>> = OnceLock::new();
+    if destruct {
+        SPAN_TX.take();
+    }
+
+    addr_of!(SPAN_TX)
+}
+
+// Message type to send on the channel. Either a span or a shutdown message for
+// graceful termination of the tokio runtime.
+enum SpanMessage {
+    Span {
+        s: Arc<BoxedSpan>
+    },
+    Shutdown,
+}
+
+#[no_mangle]
+unsafe fn nxt_otel_rs_is_init() -> u8 {
+    (*nxt_otel_rs_span_tx(false)).get().map_or(0, |_| 1)
+}
+
+#[no_mangle]
+unsafe fn nxt_otel_rs_uninit() {
+    if nxt_otel_rs_is_init() == 1 {
+        nxt_otel_rs_shutdown_tracer();
+        nxt_otel_rs_span_tx(true);
+    }
+}
+
+// potentially returns an error message
+#[no_mangle]
+unsafe fn nxt_otel_rs_init(
+    log_callback: unsafe extern "C" fn(*const c_char),
```

Comment:
I would rather this passed the log_level, rather than hard coding LOG_LEVEL_ERROR...